File: src/format.rs
use chrono::{DateTime, Duration, Local, Utc};

pub fn parse_duration(duration_str: &str) -> Result<Duration, String> {
    if let Some(hours_str) = duration_str.strip_suffix('h') {
        if let Ok(hours) = hours_str.parse::<i64>() {
            return Ok(Duration::hours(hours));
        }
    } else if let Some(days_str) = duration_str.strip_suffix('d') {
        if let Ok(days) = days_str.parse::<i64>() {
            return Ok(Duration::days(days));
        }
    } else if let Some(weeks_str) = duration_str.strip_suffix('w') {
        if let Ok(weeks) = weeks_str.parse::<i64>() {
            return Ok(Duration::weeks(weeks));
        }
    } else if let Some(months_str) = duration_str.strip_suffix('m') {
        if let Ok(months) = months_str.parse::<i64>() {
            return Ok(Duration::days(months * 30));
        }
    }

    Err(format!(
        "Invalid duration format: '{}'. Use a number followed by h (hours), d (days), w (weeks), or m (months). Examples: 24h, 7d, 2w, 3m",
        duration_str
    ))
}

pub fn format_duration(duration: Duration) -> String {
    let total_seconds = duration.num_seconds().abs();
    let sign = if duration.num_seconds() < 0 { "-" } else { "" };

    if total_seconds < 60 {
        // Show seconds to hundredths
        let total_milliseconds = duration.num_milliseconds().abs();
        let seconds = total_milliseconds / 1000;
        let hundredths = (total_milliseconds % 1000) / 10;
        format!("{}{:.0}.{:02}s", sign, seconds, hundredths)
    } else if total_seconds < 3600 {
        // Show XmYs
        let minutes = total_seconds / 60;
        let seconds = total_seconds % 60;
        format!("{}{}m{}s", sign, minutes, seconds)
    } else {
        let total_minutes = total_seconds / 60;
        let days = total_minutes / (24 * 60);
        let hours = (total_minutes % (24 * 60)) / 60;
        let minutes = total_minutes % 60;

        if days > 0 {
            // If more than a day, don't show minutes
            format!("{}{}d{}h", sign, days, hours)
        } else {
            format!("{}{}h{}m", sign, hours, minutes)
        }
    }
}

/// Parse an optional RFC3339 string into an optional `DateTime<Utc>`, silently discarding errors.
pub fn parse_rfc3339_opt(s: Option<String>) -> Option<DateTime<Utc>> {
    s.as_deref()
        .and_then(|s| DateTime::parse_from_rfc3339(s).ok())
        .map(|dt| dt.with_timezone(&Utc))
}

/// Format a DateTime<Utc> to "%Y-%m-%d %H:%M:%S" in the local timezone
pub fn format_datetime(datetime: &DateTime<Utc>) -> String {
    let local_datetime = datetime.with_timezone(&Local);
    local_datetime.format("%Y-%m-%d %H:%M:%S").to_string()
}

File: internal/format/format.go
// Package format provides duration parsing/formatting and timestamp helpers
// that mirror the original Rust implementation's behavior exactly.
package format

import (
	"fmt"
	"strconv"
	"strings"
	"time"
)

// rfc3339Layout renders timestamps in UTC with a "+00:00" offset and
// trailing-zero-trimmed fractional seconds, closely matching chrono's
// to_rfc3339() output so existing databases stay byte-compatible.
const rfc3339Layout = "2006-01-02T15:04:05.999999999-07:00"

// ParseDuration parses a human-readable duration string into a time.Duration.
//
// Accepted suffixes: h (hours), d (days), w (weeks), m (months = 30 days).
// A bare number, an unknown suffix, or an unparseable value is an error.
func ParseDuration(durationStr string) (time.Duration, error) {
	if s, ok := strings.CutSuffix(durationStr, "h"); ok {
		if hours, err := strconv.ParseInt(s, 10, 64); err == nil {
			return time.Duration(hours) * time.Hour, nil
		}
	} else if s, ok := strings.CutSuffix(durationStr, "d"); ok {
		if days, err := strconv.ParseInt(s, 10, 64); err == nil {
			return time.Duration(days) * 24 * time.Hour, nil
		}
	} else if s, ok := strings.CutSuffix(durationStr, "w"); ok {
		if weeks, err := strconv.ParseInt(s, 10, 64); err == nil {
			return time.Duration(weeks) * 7 * 24 * time.Hour, nil
		}
	} else if s, ok := strings.CutSuffix(durationStr, "m"); ok {
		if months, err := strconv.ParseInt(s, 10, 64); err == nil {
			return time.Duration(months*30) * 24 * time.Hour, nil
		}
	}

	return 0, fmt.Errorf(
		"Invalid duration format: '%s'. Use a number followed by h (hours), d (days), w (weeks), or m (months). Examples: 24h, 7d, 2w, 3m",
		durationStr,
	)
}

// FormatDuration renders a duration as a compact human-readable string.
//
//	< 1m      -> "S.HHs"  (seconds to hundredths)
//	< 1h      -> "MmSs"
//	>= 1d     -> "DdHh"   (minutes omitted)
//	otherwise -> "HhMm"
func FormatDuration(d time.Duration) string {
	ms := d.Milliseconds()
	sign := ""
	if ms < 0 {
		sign = "-"
	}
	absMs := ms
	if absMs < 0 {
		absMs = -absMs
	}

	totalSeconds := absMs / 1000

	if totalSeconds < 60 {
		seconds := absMs / 1000
		hundredths := (absMs % 1000) / 10
		return fmt.Sprintf("%s%d.%02ds", sign, seconds, hundredths)
	} else if totalSeconds < 3600 {
		minutes := totalSeconds / 60
		seconds := totalSeconds % 60
		return fmt.Sprintf("%s%dm%ds", sign, minutes, seconds)
	}

	totalMinutes := totalSeconds / 60
	days := totalMinutes / (24 * 60)
	hours := (totalMinutes % (24 * 60)) / 60
	minutes := totalMinutes % 60

	if days > 0 {
		return fmt.Sprintf("%s%dd%dh", sign, days, hours)
	}
	return fmt.Sprintf("%s%dh%dm", sign, hours, minutes)
}

// FormatRFC3339 formats a time as an RFC3339 string in UTC, matching the
// format used when persisting timestamps to the database.
func FormatRFC3339(t time.Time) string {
	return t.UTC().Format(rfc3339Layout)
}

// ParseRFC3339Opt parses an optional RFC3339 string into a *time.Time,
// silently discarding parse errors (returns nil).
func ParseRFC3339Opt(s string) *time.Time {
	if s == "" {
		return nil
	}
	t, err := time.Parse(time.RFC3339, s)
	if err != nil {
		return nil
	}
	utc := t.UTC()
	return &utc
}

// FormatDatetime formats a UTC time as "YYYY-MM-DD HH:MM:SS" in the local
// timezone.
func FormatDatetime(t time.Time) string {
	return t.Local().Format("2006-01-02 15:04:05")
}
